File: Advanced/Python/problem_22.py
class Node:
	
	def __init__(self):
		
		self.key = 0
		self.left = None
		self.right = None

def newNode(key):
	
	node = Node()
	node.key = key
	node.left = node.right = None
	return (node)

def flatten(root):


	if (root == None or root.left == None and
						root.right == None):
		return

	if (root.left != None):

		flatten(root.left)

		tmpRight = root.right
		root.right = root.left
		root.left = None

		t = root.right
		while (t.right != None):
			t = t.right


		t.right = tmpRight

	flatten(root.right)

def inorder(root):

	if (root == None):
		return
	
	inorder(root.left)
	print(root.key, end = ' ')
	inorder(root.right)


if __name__=='__main__':
	
	''' 1
		/ \
	2	 5
	/ \	 \
	3 4	 6 '''
	root = newNode(1)
	root.left = newNode(2)
	root.right = newNode(5)
	root.left.left = newNode(3)
	root.left.right = newNode(4)
	root.right.right = newNode(6)

	flatten(root)

	print("The Inorder traversal after "
		"flattening binary tree ",
		end = '')
	inorder(root)

File: Advanced/Java/problem_22.java
class Node {
	int data;
	Node left, right;
	Node(int key)
	{
		data = key;
		left = right = null;
	}
}

class BinaryTree {

	Node root;

	public void flatten(Node node)
	{
		if (node == null)
			return;
		if (node.left == null && node.right == null)
			return;
		if (node.left != null) {
	
			flatten(node.left);
	
			Node tempNode = node.right;
			node.right = node.left;
			node.left = null;
			Node curr = node.right;
			while (curr.right != null)
				curr = curr.right;
			// Insert the stored value
			curr.right = tempNode;
		}
		flatten(node.right);
	}
	public void inOrder(Node node)
	{
		if (node == null)
			return;
		inOrder(node.left);
		System.out.print(node.data + " ");
		inOrder(node.right);
	}

	public static void main(String[] args)
	{
		BinaryTree tree = new BinaryTree();

		tree.root = new Node(1);
		tree.root.left = new Node(2);
		tree.root.right = new Node(5);
		tree.root.left.left = new Node(3);
		tree.root.left.right = new Node(4);
		tree.root.right.right = new Node(6);

		System.out.println(
			"The Inorder traversal after flattening binary tree ");
		tree.flatten(tree.root);
		tree.inOrder(tree.root);
	}
}

